src/core/search/search.cc

```diff
@@ -25,65 +27,142 @@ AstExpr ParseQuery(std::string_view query) {
   return driver.Take();
 }
 
-// Add results from matched to current, keep sorted and remove duplicates.
-void UnifyResults(vector<DocId>&& matched, vector<DocId>* current, vector<DocId>* tmp) {
-  swap(*current, *tmp);  // Move current to tmp so we can directly write into current
-  current->clear();
-  current->reserve(matched.size() + tmp->size());
+// Represents an either owned or non-owned result set that can be accessed transparently.
+struct IndexResult {
+  using DocVec = vector<DocId>;
+
+  IndexResult() : value{DocVec{}} {};
+
+  IndexResult(const DocVec* dv) : value{dv} {
+    if (dv == nullptr)
+      value = DocVec{};
+  }
+
+  IndexResult(DocVec&& dv) : value{move(dv)} {
+  }
+
+  // Transparent const access to underlying value
+  const DocVec* operator->() const {
+    return holds_alternative<DocVec>(value) ? &get<DocVec>(value) : get<const DocVec*>(value);
+  }
+
+  // Move out of owned or copy borrowed
+  DocVec Take() {
+    if (holds_alternative<DocVec>(value))
+      return move(get<DocVec>(value));
+    return *get<const DocVec*>(value);
+  }
+
+  // Move out of owned or return empty
+  DocVec TakeOwnedOrEmpty() {
+    if (holds_alternative<DocVec>(value))
+      return move(get<DocVec>(value));
+    return DocVec{};
+  }
+
+  static bool BySize(const IndexResult& l, const IndexResult& r) {
```

Comment:
Having this here as a standalone function, with an out-of-context name is somewhat confusing. Maybe use a lambda below?